price-estimator/src/models/currency_pair.rs

```diff
@@ -0,0 +1,120 @@
+//! Module containing currency pair model implementation.
+
+use anyhow::{anyhow, bail, Error, Result};
+use core::token_info::TokenInfoFetching;
+use ethcontract::Address;
+use pricegraph::{Market, TokenId};
+use std::str::FromStr;
+
+/// A currency pair of two exchange tokens.
+#[derive(Clone, Debug, PartialEq)]
+pub struct CurrencyPair {
+    pub base: TokenRef,
+    pub quote: TokenRef,
+}
+
+impl CurrencyPair {
+    /// Convert the token pair into a market for `pricegraph` consumption.
+    pub async fn as_market(&self, token_infos: &dyn TokenInfoFetching) -> Result<Market> {
+        let (base, quote) = futures::try_join!(
+            self.base.as_token_id(token_infos),
+            self.quote.as_token_id(token_infos),
+        )?;
+
+        Ok(Market { base, quote })
+    }
+}
+
+impl FromStr for CurrencyPair {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        let (base, quote) = {
+            let mut parts = s.split('-');
+            let pair = parts
+                .next()
+                .and_then(|base| parts.next().map(|quote| (base, quote)));
+            if parts.next().is_none() {
+                pair
+            } else {
+                None
+            }
+        }
+        .ok_or_else(|| anyhow!("currency pair expected 'X-Y' format"))?;
+
+        Ok(CurrencyPair {
+            base: base.parse()?,
+            quote: quote.parse()?,
+        })
+    }
+}
+
+/// A token reference, either by token ID, address, or symbol.
+#[derive(Clone, Debug, PartialEq)]
+pub enum TokenRef {
+    Id(TokenId),
+    Address(Address),
+    Symbol(String),
+}
```

Comment:
One problem I see now is that it is possible for a token symbol to also be a valid id or address. We should make sure that Id has the highest priority, then Address then Symbol. In this PR it is already correct but I'd add a comment there that the order of deserialization attempts matters.

Reply:
Good observation. I think this behaviour is already guaranteed/documented by the unit tests (but a comment can't hurt).

Reply:
Added a comment in the `FromStr` implementation.